Src/AutoFixture/Kernel/CompositeSpecimenBuilder.cs

```diff
@@ -103,22 +103,60 @@ System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
 
         internal static ISpecimenBuilder ComposeIfMultiple(IEnumerable<ISpecimenBuilder> builders)
         {
-            var isSingle = builders.Take(2).Count() == 1;
-            if (isSingle)
-                return builders.Single();
+            // This code is called very frequently during graph mutation (e.g. customization).
+            // It's optimized to not perform unnecessary allocations and do not enumerate builders more than once.
+            // Code is written expecting that single builder is the most common case.
+            // See more detail here: https://github.com/AutoFixture/AutoFixture/pull/793
+            
+            ISpecimenBuilder singleItem = null;
+            List<ISpecimenBuilder> multipleItems = null;
+            bool hasItems = false;
 
-            return new CompositeSpecimenBuilder(builders);
+            using (var enumerator = builders.GetEnumerator())
+            {
+                if (enumerator.MoveNext())
+                {
+                    singleItem = enumerator.Current;
+                    hasItems = true;
+
+                    while (enumerator.MoveNext())
+                    {
+                        // We are here only if more than one item is present.
+                        // When we create collection for multiple items, we add the already enumerated one as well.
+                        if (multipleItems == null)
+                        {
+                            multipleItems = new List<ISpecimenBuilder> {singleItem};
+                        }
+                        
+                        multipleItems.Add(enumerator.Current);
+                    }
+                }
+            }
+
+            if (!hasItems)
+            {
+                //It's very rare case and it does't make sense to optimize it

```

Comment:
Nitpick: We usually add a space after `//` in code comments, and a period (`.`) at the end of each sentence.